src/libcore/mem.rs

```diff
@@ -1002,6 +1002,26 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
 ///
 /// [`Vec<T>`]: ../../std/vec/struct.Vec.html
 ///
+/// # Layout
+///
+/// `MaybeUninit<T>` is guaranteed to have the same size and alignment as `T`:
+///
+/// ```rust
+/// use std::mem::{MaybeUninit, size_of, align_of};
+/// assert_eq!(size_of::<MaybeUninit<u64>>(), size_of::<u64>());
+/// assert_eq!(align_of::<MaybeUninit<u64>>(), size_of::<u64>());
+/// ```
+///
+/// However remember that a type *containing* a `MaybeUninit<T>` is not necessarily the same
+/// layout.  In particular, because any bit value is valid, the compiler can't apply
+/// non-zero/niche-filling optimizations:
+///
+/// ```rust
+/// # use std::mem::{MaybeUninit, size_of, align_of};
+/// assert_eq!(size_of::<Option<bool>>(), 1);
+/// assert_eq!(size_of::<Option<MaybeUninit<bool>>>(), 2);
+/// ```
+///
```

Comment:
Could you move this down below examples? They are closely connected to the previous story about initialization invariants.